### PR TITLE
update: Read levels from file

### PR DESCRIPTION
**Pull request by mrshu**
- Add ability to load levels from file, so that they do not have to be
  hardcoded.
- Provide a sample `.gta` file which shows how can these levels be
  structured, prepared and executed.
- Improve README in order to explain in greater detail how the
  `.gta` files work and how can they be used to achieve great things.
- Add ability to just print (not execute) loaded levels in order to be
  able to check whether the formatting is correct/as expected.

Signed-off-by: mr.Shu mr@shu.io

### CODE REVIEW

**Comment by pepol** [comment]
LGTM

**Comment by mrshu** [comment]
@pepol @Foxiy Please do take a closer look, it should be clean, the tests are not happening at this stage...

File: sample_challenge.gta [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	$ cd /usr
+
+Got it?
+
+I am pretty sure you do. Now use this cd comand and go to the /tmp directory
+
+--------------------
+
+name: l01
+test: test $(pwd) = "/tmp"
```
Comment (vidriduch) [inline sample_challenge.gta:24]: @mrshu Text says that I should cd to /usr but test is written for /tmp

File: main.go [hunk 2]
```diff
@@ -2,61 +2,31 @@ package main
...
-	})
-
-	challenge.AddLevel(levels.Level{
-		Name:    `l01`,
-		TestCmd: `test $(pwd) = "/tmp"`,
-		Text: `
-I see you made it, awesome !
+	if len(flag.Args()) < 1 {
+		fmt.Printf("\n\nNo input file\n\n")
+		fmt.Printf("usage: %s path\n\n", flag.Args()[0])
```
Comment (vidriduch) [inline main.go]: @mrshu When no argument is specified this line throws `panic: runtime error: index out of range`. I think flag doesn't have have 0th element name of the file, which would make sense since you are checking if a length of `flag.Args()` is 0.

File: sample_challenge.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+PROMPT_COMMAND="$(pwd)/go-term-adventure $(pwd)/sample_challenge.gta" HISTFILE=$HOME/.advhistory bash
```
Comment (vidriduch) [inline sample_challenge.sh]: @mrshu If this should be a `sample file` shouldn't this also contain option to specify --print?

File: sample_challenge.gta (hunk 1 above)
Comment (vidriduch) [inline sample_challenge.gta:24]: @mrshu ignore this I misread the instruction. Sorry

File: sample_challenge.sh (hunk 3 above)
Comment (mrshu) [inline sample_challenge.sh]: I do not think I follow, can you rephrase that?

Comment (vidriduch) [inline sample_challenge.sh]: I just mean if there shouldn't be  
`PROMPT_COMMAND="$(pwd)/go-term-adventure $1 $(pwd)/sample_challenge.gta"`  
or something

Comment (mrshu) [inline sample_challenge.sh]: @Foxiy I see. That will have to wait for another comment which will transform this into some sort of a template, that will take care of most of these things and only require you to specify the file that specifies the respective levels.

It short: it's outside of the scope of this PR.

File: main.go (hunk 2 above)
Comment (mrshu) [inline main.go]: @Foxiy Thanks, this is indeed a bug which should be fixed now.

Comment (mrshu) [inline main.go]: @Foxiy Thanks, this is indeed a bug which should be fixed now.

**Comment by mrshu** [comment]
Thanks for the review @Foxiy, this is ready for another pass and can be most probably directly merged.

File: sample_challenge.gta [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+First of all, you need to learn how to move from one directory to another. This
+can be done using the "cd" commad which is a very short abbreviation for
+**'change directory'**.  For instance if I wanted to go to the /usr directory
+I would type
+
+	$ cd /usr
+
+Got it?
+
+I am pretty sure you do. Now use this cd comand and go to the /tmp directory
```
Comment (vidriduch) [inline sample_challenge.gta]: @mrshu Pretty sure `comand` should be `command`... just a simple typo

File: sample_challenge.gta [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+# First alternative bash excercise
+
+Hi there.
+
+I understand that a command line talking to you might not make sense right now
+but bear with me, things *will* get better.
+
+First of all, you need to learn how to move from one directory to another. This
+can be done using the "cd" commad which is a very short abbreviation for
```
Comment (vidriduch) [inline sample_challenge.gta]: @mrshu Also simple typo `commad` should probably be `command`

**Comment by vidriduch** [comment]
@mrshu This is fine now I couldn't find anything else except those two typos so if @pepol is good with this we can merge it.

**Comment by pepol** [comment]
@mrshu as @Foxiy noted, please fix those two typos, otherwise LGTM.

However I'd like to question the motive behind using level descriptors as plaintext documents. I thought the whole idea of this being written in compiled language was to prevent level data itself being (more) easily manipulated. Was I wrong in this belief? And if not, how do you plan to address this?

**Comment by mrshu** [comment]
@pepol Building new levels by compiling an updated version of the main `.go` file is highly impractical by any measure you choose. I completely understand where you are coming from but the real reason (at least for me) for this being written in Go was a static binary that is the output of the build process.

That being said, I am definitely open to being able to read encrypted versions of these documents. That will of course complicate matters a bit, but in the long term it provide at least the same level of "secrecy" of these documents as if they were embedded in some sort of a binary.

I do not think there is any way we can really fight level data manipulation: we can only make it more difficult for an "attacker" to get to that data and figure out what its meaning is. That is essentially how I plan to address this.
